app/bundles/LeadBundle/Tests/Entity/TagRepositoryTest.php

```diff
@@ -0,0 +1,113 @@
+{
+    public function testSaveEntityWithUniqueNewTag()
+    {
+        $entity = new Tag('sometag');
+
+        $mockEntityManager = $this->getMockBuilder(EntityManager::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['persist', 'flush'])
+            ->getMock();
+
+        $mockEntityManager->expects($this->once())
+            ->method('persist')
+            ->with($entity);
+
+        $mockRepository = $this->getMockBuilder(TagRepository::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['getEntityManager', 'findOneBy'])
+            ->getMock();
+
+        $mockRepository->method('getEntityManager')
+            ->willReturn($mockEntityManager);
+
+        $mockRepository->expects($this->once())
+            ->method('findOneBy')
+            ->with(['tag' => 'sometag'])
+            ->willReturn(null);
+
+        $mockRepository->saveEntity($entity);
+
+        $this->assertNull($entity->getId());
+    }
+
+    public function testSaveEntityWithExistingTag()
+    {
+        $entity = new Tag('sometag');
+        $entity->setId(45);
+
+        $mockEntityManager = $this->getMockBuilder(EntityManager::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['persist', 'flush'])
+            ->getMock();
+
+        $mockEntityManager->expects($this->once())
+            ->method('persist')
+            ->with($entity);
+
+        $mockRepository = $this->getMockBuilder(TagRepository::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['getEntityManager', 'findOneBy'])
+            ->getMock();
+
+        $mockRepository->method('getEntityManager')
+            ->willReturn($mockEntityManager);
+
+        $mockRepository->expects($this->never())
+            ->method('findOneBy');
+
+        $mockRepository->saveEntity($entity);
+
+        $this->assertSame(45, $entity->getId());
+    }
+
+    public function testSaveEntityWithNewDuplicatedTag()
+    {
+        $entity = new Tag('sometag');
+
+        $mockEntityManager = $this->getMockBuilder(EntityManager::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['persist', 'flush'])
+            ->getMock();
+
+        $mockEntityManager->expects($this->never())
+            ->method('persist');
+
+        $mockRepository = $this->getMockBuilder(TagRepository::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['getEntityManager', 'findOneBy'])
+            ->getMock();
+
+        $mockRepository->method('getEntityManager')
+            ->willReturn($mockEntityManager);
+
+        $existingTag = new Tag('sometag');
+        $existingTag->setId(23);
+
+        $mockRepository->expects($this->once())
+            ->method('findOneBy')
+            ->with(['tag' => 'sometag'])
+            ->willReturn($existingTag);
+
+        $mockRepository->saveEntity($entity);
+
+        $this->assertSame(23, $entity->getId());
```

Comment:
This behaviour is very confusing. I have a new entity, which gets same ID like other entity. If I persist it and call flush, DB will assign a new ID, so I have 2 entities in my DB. No unique key is on an entity, so there will not be any exception thrown.

Reply:
I don't understand your comment. There is no exception in it.

Reply:
Ok, I understand the code now (many comments above are related to this piece of code).
This is very badly designed - why to create a new entity and check if it exists later when you want to save it? Current workflow is:
1) Select an entity form DB. Create a new (not persisted) entity if it does not exists.
2) Do some stuff with it
3) Call save
4) Check if it exists - if yes, set ID of object to an ID of entity from DB and do not save.

So you have an object, which is a copy of entity, but not entity itself. It is not managed by EntityManager, which could be a problem - you will get an Exception if you use it in another entity (`$someEntity->setTag($tag); $em->persist($someEntity); $em->flush();`

Another problem with current approach could be if `$someEntity` is configured with cascade persist. In this case Doctrine inserts this new entity to the database. There is no unique key on the tag property, so there will be no error and 2 tags with same string will be stored.

Would be much better if:
1) Select entity from DB according to conditions.
2) Create it if it does not exists.
3) Save (= insert or update, but Doctrine will take care about this).

This approach is much more simple (do not need a `setId` method) and with no side effects.

Reply:
Here is my reasoning:

The new tags are created at several places without knowing if they exist or not. For example:
https://github.com/mautic/mautic/pull/5204/files#diff-effadf31fa0a0906daae388dc7dc4999R652 or https://github.com/mautic/mautic/pull/5204/files#diff-74cc87987a53404385f0f0eaf2ae4acdR2084
So instead of `new Tag($tag)` you suggest there should be something like `$repository->$this->findOneBy(['tag' => $tag])`, right? Instead of calling it at one place, it will be called on many. But I can do that.

Another reason was to avoid unnecessary INSERT query if there is no change that is needed to be inserted whatsoever. So you suggest to fire the query either way?

Another reason is that I'll have to make the TagApiController much more complicated because it will have to ensure the duplication won't happen during edit, new, batch edit and batch insert methods.

Reply:
1) Ok, but that suggests this is a code smell. Why to have such logic in Controller? You need to find if it exists sooner or later anyway. This piece of code cannot work (or I consider it as a Doctrine bug, if it is working now:
https://github.com/mautic/mautic/pull/5204/files#diff-74cc87987a53404385f0f0eaf2ae4acdR2084)
If you do not persist this tag to DB, you will not be able to save a Lead or Tag will not be added (because of copy problem I mentioned before). And if you persist, you will get a same tag in DB twice.
2) I would put it to separate model class - something like `getOrCreateTageByString()`. This ensures that you have valid entity all time.
3) I do not understand your comment about unnecessary INSERT - there will be no tag for Lead if INSERT does not happen. Did you mean update?

Reply:
1. We are getting away from what this PR is about again. I don't intent to rewrite the tag functionality with this PR. I just need to implement API endpoints for it and make sure the code changes work with the code which is there already. The code you don't like and I'm not fun of it either is the simplest solution. It will complicate code on many places if we'd do it the way you want it. If not then maybe I don't understand your suggestions clearly.
2. TagModel is not a good place for such method?
3. Yep, update.

Reply:
1) I do not agree because I think this code just will not work and is potential buggy. What happens if you call `$lead->addTag(new Tag($tag));` and you do not persist and save this tag later?
I do not like hack according to simple solution. If you do not want to rewrite old code, you can always create new methods. But I think little refactoring is worth it.
2) Could be
3) You can skip update if you want (`if ($this->entityManager->contains($tag)) {return;}`). But I think Doctrine will not fire update if there is no change on the entity anyway.

Reply:
The Lead/Tags association has cascade persist so if a new tag is added, and the Lead entity is saved, it'll persist. Currently behavior is based on tags being saved through the Lead entity so we can capture changes to know when to rebuild a segment based on tags for cloud Mautic.

Reply:
Where is the code responsible for catching duplicates If it has cascade persist?
If you call:
```
$lead1->addTag(new Tag('tag string'));
$lead2->addTag(new Tag('tag string'));
```
I am pretty sure that the result will be 2 same Tag entites in DB.
Or did I missed anything?